src/dash-table/derived/edges/type.ts

```diff
@@ -0,0 +1,197 @@
+import * as R from 'ramda';
+import { CSSProperties } from 'react';
+
+import { OptionalMap, OptionalProp, PropOf } from 'core/type';
+import py2jsCssProperties from '../style/py2jsCssProperties';
+
+export type Edge = any;
+
+type BorderProp =
+    PropOf<CSSProperties, 'borderBottom'> |
+    PropOf<CSSProperties, 'borderLeft'> |
+    PropOf<CSSProperties, 'borderRight'> |
+    PropOf<CSSProperties, 'borderTop'>;
+
+export type BorderStyle =
+    OptionalMap<CSSProperties, 'borderBottom', [OptionalProp<CSSProperties, 'borderBottom'>, number]> &
+    OptionalMap<CSSProperties, 'borderLeft', [OptionalProp<CSSProperties, 'borderLeft'>, number]> &
+    OptionalMap<CSSProperties, 'borderRight', [OptionalProp<CSSProperties, 'borderRight'>, number]> &
+    OptionalMap<CSSProperties, 'borderTop', [OptionalProp<CSSProperties, 'borderTop'>, number]>;
+
+export const BORDER_PROPERTIES: BorderProp[] = [
+    'borderBottom',
+    'borderLeft',
+    'borderRight',
+    'borderTop'
+];
+
+export const BORDER_PROPERTIES_AND_FRAGMENTS: string[] = R.uniq(
+    R.filter(
+        p => p.indexOf('border') === 0,
+        Array.from(py2jsCssProperties.values())
+    )
+);
+
+export interface IEdgesMatrix {
+    getEdge(i: number, j: number): Edge;
+    getEdges(): Edge[][];
+    getWeight(i: number, j: number): number;
+    isDefault(i: number, j: number): boolean;
+}
+
+export interface IEdgesMatrices {
+    getEdges(): {
+        horizontal: Edge[][],
+        vertical: Edge[][]
+    };
+    getMatrices(): {
+        horizontal: EdgesMatrix,
+        vertical: EdgesMatrix
+    };
+    getStyle(i: number, j: number): CSSProperties;
+}
+
+export class EdgesMatrix implements IEdgesMatrix {
+    private weights: number[][];
+    private edges: Edge[][];
+
+    public readonly rows: number;
+    public readonly columns: number;
+    public readonly defaultEdge: Edge | undefined;
+
+    constructor(m: EdgesMatrix);
+    constructor(
+        rows: number,
+        columns: number,
+        defaultEdge?: Edge
+    );
+    constructor(
+        rowsOrMatrix: number | EdgesMatrix,
+        columns?: number,
+        defaultEdge?: Edge
+    ) {
```

Comment:
Fwiw, this is how a ctor is overloaded in typescript -- define n ctors with their own definitions and then declare one that encompasses them all. As TS generates absolutely no new code, the developer is responsible for making it all work.